Add nxp edac driver

yvesll opened the pull request:
Added NXP's EDAC implementation based on EIM and ERM

decsny reviewed (commented):
Looks like a very good first PR.  But can you explain why this Devicetree structure with 4(+?) nodes was chosen?

yvesll commented:
> Looks like a very good first PR. But can you explain why this Devicetree structure with 4(+?) nodes was chosen?

The edac driver requires two peripherals: erm for error reporting and eim for error injection. Eim has many channels which represent different memory regions.

Holt-Sun commented on drivers/edac/edac_nxp.c:
```diff
@@ -0,0 +1,345 @@
+		break;
+	default:
+		LOG_ERR("No error type found.");
+		return 0;
```
return -EINVAL;

Holt-Sun commented on drivers/edac/edac_nxp.c:
```diff
@@ -0,0 +1,345 @@
+	const struct edac_nxp_config *config = dev->config;
+	struct edac_nxp_data *data = dev->data;
+
+	return ERM_GetErrorCount(config->erm_base, data->erm_channel);
```
return (int)ERM_GetErrorCount(config->erm_base, data->erm_channel);

Holt-Sun commented on drivers/edac/edac_nxp.c:
```diff
@@ -0,0 +1,345 @@
+		return -EINVAL;
+	}
+	data->erm_channel = CONFIG_EDAC_NXP_ERM_DEFAULT_CHANNEL;
+	config->irq_config_func(dev);
```
```suggestion
    /* Clear any latched status before enabling interrupts */
    ERM_ClearInterruptStatus(config->erm_base, data->erm_channel, kERM_AllInterruptsEnable);

    config->irq_config_func(dev);
    ERM_EnableInterrupts(config->erm_base, data->erm_channel, kERM_AllInterruptsEnable);
```

Holt-Sun commented:
> Quality Gate passed

Please fix the issue in "Quality Gate passed".

decsny commented on drivers/edac/Kconfig.nxp:
would be good to name this file like Kconfig.nxp_edac , in case there is some other hardwares from nxp later, we have one kconfig file per device driver usually

decsny commented on drivers/edac/edac_nxp.c:
```diff
@@ -0,0 +1,348 @@
+	uint32_t erm_channel;
+};
+
+#ifdef CONFIG_SOC_FAMILY_MCXN
+static void enable_ecc(uint32_t mask)
+{
+	/* Enable memory channel RAM ECC functionality. */
+	SYSCON->ECC_ENABLE_CTRL = mask;
+}
+#elif defined(CONFIG_SOC_FAMILY_MCXA)
+static void enable_ecc(uint32_t mask)
+{
+	/* Enable memory channel RAM ECC functionality. */
+	SYSCON->RAM_CTRL = mask;
+}
+#else
+static void enable_ecc(uint32_t mask)
+{
+}
+#endif
```
I suggest to put this in soc.h instead of ifdeffing here

decsny commented on drivers/edac/edac_nxp.c:
```diff
@@ -0,0 +1,348 @@
+};
+
+/* NOTE to checkpatch error: SPACING: need consistent spacing around '*' (ctx:WxV)  */
+#define EIM_CONFIG                                                                                 \
+	EIM_Type *eim_base;                                                                        \
+	const struct edac_nxp_eim_channel *eim_channels;                                           \
+	uint8_t eim_channel_num;
+#else
+#define EIM_CONFIG
+#endif /* CONFIG_EDAC_ERROR_INJECT */
```
this doesn't need to be a macro, just #ifdef within the config struct definition, its actually less confusing that way

decsny replied:
@keith-zephyr I think the compliance script is doing something wrong here, I saw you are coding guideline maintainer, can you help figuring out what is the problem here? Because this code in the PR now is weird due to trying to get around this script, I guess

decsny commented on drivers/edac/edac_nxp.c:
```diff
@@ -0,0 +1,348 @@
+{
+	const struct edac_nxp_config *config = dev->config;
+	struct edac_nxp_data *data = dev->data;
+#ifdef CONFIG_EDAC_ERROR_INJECT
+	EIM_Init(config->eim_base);
+	EIM_EnableGlobalErrorInjection(config->eim_base, true);
+	LOG_INF("EIM driver initialized");
+#endif /* CONFIG_EDAC_ERROR_INJECT */
```
nit is to put blank lines before and after this block

yvesll commented:
> @yvesll you might need to rebase to get past some twister bug

Sure, I will do this together with the support for MCXE31

yvesll committed:
drivers: edac: Add NXP EDAC driver

Add edac driver for NXP's ERM and EIM peripherals. It can inject ECC
error to specific channel within EIM and then report the error address,
syndrome and count within ERM.

Signed-off-by: Yves Wang <zhengjia.wang@nxp.com>

yvesll committed:
dts: nxp: add edac support for MCX platforms

Add eim, erm and edac instance for frdm_mcxa153, frdm_mcxn236,
frdm_mcxn947, frdm_mcxe247 and frdm_mcxe31b.

Signed-off-by: Yves Wang <zhengjia.wang@nxp.com>

yvesll committed:
samples: edac: update the app to support more platforms

Decouple the edac sample with IBECC.

Signed-off-by: Yves Wang <zhengjia.wang@nxp.com>

yvesll committed:
doc: Update edac shell sample's readme

Update the console log in readme to make it aligned with the actual
output

Signed-off-by: Yves Wang <zhengjia.wang@nxp.com>

Holt-Sun commented on include/zephyr/drivers/edac/edac_mcux_erm.h:
```diff
@@ -0,0 +1,21 @@
+/* Callback data provided to function passed to notify_cb_set */
+struct edac_nxp_callback_data {
+	/* Number of corrected errors */
+	uint8_t corr_err_count;
```
If the corr_err_count could  exceed 255, then should use uint32_t.

yvesll replied:
It will not, ERM only use 8 bit for error counter.

kartben reviewed (dismissed):
+1 for docs. Thank you!

yvesll commented:
ci check may blocked by #100426
